src/main/scala/mesosphere/marathon/core/task/jobs/impl/OverdueTasksActor.scala

```diff
@@ -81,51 +75,317 @@ private[jobs] object OverdueTasksActor {
+
   }
 
   private[jobs] case class Check(maybeAck: Option[ActorRef])
+
+  /**
+    * Create the Overdue tasks graph
+    *
+    * @param support
+    * @param checkTick
+    * @param reconciliationTick
+    * @param reconciliationStatusUpdates
+    * @param ec
+    * @param mat
+    * @return
+    */
+  private[impl] def overdueTasksGraph(
+    support: Support,
+    checkTick: Source[Instant, Cancellable],
+    reconciliationTick: Source[Instant, Cancellable],
+    maxReconciliations: Int,
+    bufferSize: Int = 100,
+    reconciliationStatusUpdates: Source[ReconciliationStatusUpdate, NotUsed])(implicit ec: ExecutionContext, mat: Materializer, system: ActorSystem): Graph[ClosedShape, NotUsed] =
+    GraphDSL.create() {
+      implicit builder =>
+        import GraphDSL.Implicits._
+
+        //
+        // [instancesBySpec] - Returns all the instances in marathon
+        //
+        val instancesBySpec = Flow[Instant]
+          .mapAsync(1) { instant =>
+            support.instancesBySpec().map(_.allInstances -> instant)
+          }
+
+        //
+        // [timeoutOverdueReservations] - Times out
+        //
+        val timeoutOverdueReservations = Flow[(Seq[Instance], Instant)]
+          .mapAsync(1){
+            case (instances, instant) => support.timeoutOverdueReservations(Timestamp(instant), instances)
+          }
+          .to(Sink.ignore)
+
+        //
+        // [overdueInstancesFilter] - Filter out
+        //
+        val overdueInstancesFilter: Flow[(Seq[Instance], Instant), Instance, NotUsed] = Flow[(Seq[Instance], Instant)]
+          .mapConcat { case (instances, t) => support.filterOverdueTasks(Timestamp(t), instances) }
+
+        //
+        // [reconciliationTracker] - Keeps track of reconciliation requests and ticks and prepares kill events
+        //
+        val reconciliationTracker = builder.add(new ReconciliationTracker(
+          support.reconcileTasksFuture(_)(system.dispatchers.lookup("marathon-blocking-dispatcher")),
+          bufferSize,
+          maxReconciliations)
+        )
+
+        //
+        // [killTasks] - Kills the tasks by their instance ID
+        //
+        val killTasks: Sink[Instance, NotUsed] = Flow[Instance]
+          .mapAsync(1)(instance => support.killInstance(instance, KillReason.Overdue))
+          .to(Sink.ignore)
+
+        // (Utility)
+        val bcast = builder.add(Broadcast[(Seq[Instance], Instant)](2))
+
+        /////////////////////////////////////////////
+        //format: OFF
+                                         bcast ~> timeoutOverdueReservations
+        checkTick ~> instancesBySpec ~>  bcast ~> overdueInstancesFilter ~> reconciliationTracker.in0
+        reconciliationStatusUpdates            ~>                           reconciliationTracker.in1
+        reconciliationTick                     ~>                           reconciliationTracker.in2
+
+        reconciliationTracker.out ~> killTasks
+
+        //format: ON
```

Comment:
There is a fan-out, such as `instancesBySpec -> timeoutOverdueReservations` and `instancesBySpec ~> overdueInstancesFilter `. I don't think a simple flow could describe this.

Reply:
Why do we have to fan out for this? I can see the benefit but I think we should take smaller steps.

Reply:
So, you are proposing to repeat the source two times?

Reply:
I think this would be easier to tackle if we broke down this task. There will be a broadcast but it would be easier to define separated flows for overdue confirmations, launches and status updates and then use them to build a graph.